### PR TITLE
Add Repository Group resource with IAM

### PR DESCRIPTION
**Pull request by JumiDeluxe**
**Release Note Template for Downstream PRs (will be copied)**

See [Write release notes](https://googlecloudplatform.github.io/magic-modules/contribute/release-notes/) for guidance.

Add new resources for Gemini Code Assist

```release-note:new-resource
`google_gemini_repository_group` (beta only)
```
```release-note:new-resource
`google_gemini_repository_group_iam_member` (beta only)
```
```release-note:new-resource
`google_gemini_repository_group_iam_binding` (beta only)
```
```release-note:new-resource
`google_gemini_repository_group_iam_policy` (beta only)
```

### CODE REVIEW

**Comment by JumiDeluxe** [comment]
> Hi there, I'm the Modular magician. I've detected the following information about your changes:
> 
> ## Diff report
> Your PR generated some diffs in downstreams - here they are.
> 
> `google` provider: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-12424-old..auto-pr-12424) ( 10 files changed, 537 insertions(+), 2 deletions(-)) `google-beta` provider: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-12424-old..auto-pr-12424) ( 22 files changed, 3099 insertions(+), 5 deletions(-)) `terraform-google-conversion`: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-12424-old..auto-pr-12424) ( 4 files changed, 594 insertions(+))
> 
> ## Missing test report
> Your PR includes resource fields which are not covered by any test.
> 
> Resource: `google_gemini_repository_group_iam_binding` (2 total tests) Please add an acceptance test which includes these fields. The test should include the following:
> 
> ```hcl
> resource "google_gemini_repository_group_iam_binding" "primary" {
>   condition {
>     description = # value needed
>     expression  = # value needed
>     title       = # value needed
>   }
> }
> ```
> 
> Resource: `google_gemini_repository_group_iam_member` (1 total tests) Please add an acceptance test which includes these fields. The test should include the following:
> 
> ```hcl
> resource "google_gemini_repository_group_iam_member" "primary" {
>   condition {
>     description = # value needed
>     expression  = # value needed
>     title       = # value needed
>   }
> }
> ```

Just a reminder that this resource will not support conditions at this moment

File: mmv1/third_party/terraform/services/gemini/resource_gemini_repository_group_test.go.tmpl [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+  location = "us-central1"
+  connection_id = "tf-test-cloudaicompanion1"
+  disabled = false
+
+  github_config {
+    github_app = "DEVELOPER_CONNECT"
+    app_installation_id = 54180648
+
+    authorizer_credential {
+      oauth_token_secret_version = "projects/502367051001/secrets/tf-test-cloudaicompanion-github-oauthtoken-c42e5c/versions/1"
```
[the remark this answers is not in the document]
Comment (JumiDeluxe) [reply mmv1/third_party/terraform/services/gemini/resource_gemini_repository_group_test.go.tmpl:111]: Manual setup of DeveloperConnect app with repository access is needed for some GH account

File: mmv1/third_party/terraform/services/gemini/resource_gemini_repository_group_test.go.tmpl [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+  parent_connection = google_developer_connect_connection.github_conn.connection_id
+  clone_uri = "https://github.com/CC-R-github-robot/tf-test.git"
+  location = "us-central1"
+  annotations = {}
+}
+
+resource "google_developer_connect_connection" "github_conn" {
+  provider = google-beta
+  location = "us-central1"
+  connection_id = "tf-test-cloudaicompanion1"
```
[the remark this answers is not in the document]
Comment (JumiDeluxe) [reply mmv1/third_party/terraform/services/gemini/resource_gemini_repository_group_test.go.tmpl]: Randomized suffixes and changed id for the noBootstrap test

**Comment by JumiDeluxe** [comment]
Rebased onto master

File: mmv1/third_party/terraform/services/gemini/resource_gemini_repository_group_test.go.tmpl [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+				ImportStateVerify:       true,
+				ImportStateVerifyIgnore: []string{"code_repository_index_id", "labels", "location", "repository_group_id", "terraform_labels"},
+			},
+		},
+	})
+}
+
+func testAccGeminiRepositoryGroup_noBootstrap(context map[string]interface{}) string {
+	return acctest.Nprintf(`
+resource "google_gemini_code_repository_index" "cri" {
```
[the remark this answers is not in the document]
Comment (devid-shyliuk) [reply mmv1/third_party/terraform/services/gemini/resource_gemini_repository_group_test.go.tmpl:72]: move declaration of google_gemini_code_repository_index

[the remark this answers is not in the document]
Comment (JumiDeluxe) [reply mmv1/third_party/terraform/services/gemini/resource_gemini_repository_group_test.go.tmpl]: Sure, added a test with 5 RGs and moved the test function

File: mmv1/products/gemini/RepositoryGroup.yaml [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+  parent_resource_attribute: 'repository_group_id'
+  method_name_separator: ':'
+  fetch_iam_policy_verb: 'GET'
+  set_iam_policy_verb: 'POST'
+  import_format:
+    - 'projects/{{project}}/locations/{{location}}/codeRepositoryIndexes/{{code_repository_index_id}}/repositoryGroups/{{repository_group_id}}'
+    - '{{repository_group_id}}'
+  allowed_iam_role: 'roles/cloudaicompanion.repositoryGroupsUser'
+error_retry_predicates:
+  - 'transport_tpg.IsRepositoryGroupQueueError'
```
Comment (devid-shyliuk) [inline mmv1/products/gemini/RepositoryGroup.yaml:82]: guess, you would like to add the following isCodeRepositoryIndexUnreadyError here

File: mmv1/third_party/terraform/transport/error_retry_predicates.go [hunk 5]
```diff
@@ -115,6 +115,21 @@ func is409OperationInProgressError(err error) (bool, string) {
...
 
+// Code Repository Index is a long running operation
+// The resource takes time to change it's state from "CREATING" to "ACTIVE"
+func IsCodeRepositoryIndexUnreadyError(err error) (bool, string) {
+	gerr, ok := err.(*googleapi.Error)
+	if !ok {
+		return false, ""
+	}
+
+	if gerr.Code == 409 && strings.Contains(gerr.Body, "parent resource not in ready state") {
```
Comment (slevenick) [inline mmv1/third_party/terraform/transport/error_retry_predicates.go:126]: FYI when we start depending on error message contents it makes it extremely difficult to change the error messages on the API side in the future. That's the state of the world today so it's not unreasonable to add this, but it does have downsides

File: mmv1/third_party/terraform/services/gemini/iam_gemini_repository_group_test.go.tmpl [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+	"github.com/hashicorp/terraform-plugin-testing/helper/resource"
+
+	"github.com/hashicorp/terraform-provider-google/google/acctest"
+	"github.com/hashicorp/terraform-provider-google/google/envvar"
+)
+
+// To run tests locally please replace the `oauth_token_secret_version` with your secret manager version.
+// More details: https://cloud.google.com/developer-connect/docs/connect-github-repo#before_you_begin
+
+func TestAccGeminiRepositoryGroupIamBinding(t *testing.T) {
```
Comment (slevenick) [inline mmv1/third_party/terraform/services/gemini/iam_gemini_repository_group_test.go.tmpl:17]: Were you unable to get the IAM tests to generate? We shouldn't need handwritten test files like this if the IAM generation works as expected

File: mmv1/products/gemini/RepositoryGroup.yaml [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+---
+name: RepositoryGroup
+description: The resource for managing Repository Group for Gemini Code Assist.
+min_version: 'beta'
+base_url: projects/{{project}}/locations/{{location}}/codeRepositoryIndexes/{{code_repository_index_id}}/repositoryGroups
```
[the remark this answers is not in the document]
Comment (JumiDeluxe) [reply mmv1/products/gemini/RepositoryGroup.yaml:18]: Yes, I've changed it to the `code_repository_index_id` to make it clear that it should hold the same value as in the `CodeRepositoryIndex.yaml` resource. I can change it back.

File: mmv1/products/gemini/RepositoryGroup.yaml [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+    - delete
+    - update
+  type: OpAsync
+  result:
+    resource_inside_response: true
+    path: response
+  error:
+    path: error
+    message: message
+  include_project: false
```
[the remark this answers is not in the document]
Comment (JumiDeluxe) [reply mmv1/products/gemini/RepositoryGroup.yaml:70]: Surprisingly, setting it to `true` generates providers with `no new variables on left side of :=` errors in the Create, Update and Delete functions due to the `project` variable declaration

File: mmv1/third_party/terraform/services/gemini/iam_gemini_repository_group_test.go.tmpl (hunk 6 above)
Comment (JumiDeluxe) [reply mmv1/third_party/terraform/services/gemini/iam_gemini_repository_group_test.go.tmpl]: Replaced handwritten tests with generated ones

File: mmv1/products/gemini/RepositoryGroup.yaml (hunk 8 above)
[the remark this answers is not in the document]
Comment (JumiDeluxe) [reply mmv1/products/gemini/RepositoryGroup.yaml:69]: Will change the value once the bug is resolved

File: mmv1/third_party/terraform/services/gemini/iam_gemini_repository_group_test.go.tmpl (hunk 6 above)
Comment (JumiDeluxe) [reply mmv1/third_party/terraform/services/gemini/iam_gemini_repository_group_test.go.tmpl]: It seems that IAM tests that generate with `t.Parallel` still cause conflicts on the API side, so until turning it off in generating tests is possible, we will go with handwritten ones, which are the same but not parallel.

**Review by slevenick — APPROVED** [review]

**Pull request merged**